pkg/applyset/applyset.go

```diff
@@ -0,0 +1,543 @@
+
+// ID is the label value that we are using to identify this applyset.
+// Format: base64(sha256(<name>.<namespace>.<kind>.<apiVersion>)), using the URL safe encoding of RFC4648.
+func (a *applySet) ID() string {
+	unencoded := strings.Join([]string{
+		a.parent.GetName(),
+		a.parent.GetNamespace(),
+		a.parent.GetKind(),
+		a.parent.GroupVersionKind().Group,
+	}, ApplySetIDPartDelimiter)
+	hashed := sha256.Sum256([]byte(unencoded))
+	b64 := base64.RawURLEncoding.EncodeToString(hashed[:])
+	// Label values must start and end with alphanumeric values, so add a known-safe prefix and suffix.
+	return fmt.Sprintf(V1ApplySetIdFormat, b64)
+}
+
+func (a *applySet) injectToolLabels(labels map[string]string) map[string]string {
+	if labels == nil {
+		labels = make(map[string]string)
+	}
+	if a.toolLabels != nil {
+		for k, v := range a.toolLabels {
+			labels[k] = v
+		}
+	}
+	return labels
+}
+
+func (a *applySet) InjectApplysetLabels(labels map[string]string) map[string]string {
+	if labels == nil {
+		labels = make(map[string]string)
+	}
+	labels[ApplysetPartOfLabel] = a.ID()
+	return labels
+}
+
+type applySetUpdateMode string
+
+var updateToLatestSet applySetUpdateMode = "latest"
+var updateToSuperset applySetUpdateMode = "superset"
+
+// updateParentLabelsAndAnnotations updates the parent labels and annotations.
+func (a *applySet) updateParentLabelsAndAnnotations(ctx context.Context, mode applySetUpdateMode) (sets.Set[string],
+	sets.Set[string], error) {
+	original, err := meta.Accessor(a.parent.DeepCopyObject())
+	if err != nil {
+		return nil, nil, err
+	}
+
+	parentPatch := &unstructured.Unstructured{}
+	parentPatch.SetUnstructuredContent(map[string]interface{}{
+		"apiVersion": a.parent.GetAPIVersion(),
+		"kind":       a.parent.GetKind(),
+		"metadata": map[string]interface{}{
+			"name":      a.parent.GetName(),
+			"namespace": a.parent.GetNamespace(),
+		},
+	})
+	// Generate and append the desired labels to the parent labels
+	desiredLabels := a.desiredParentLabels()
+	labels := a.parent.GetLabels()
+	if labels == nil {
+		labels = make(map[string]string)
+	}
+	for k, v := range desiredLabels {
+		labels[k] = v
+	}
+	parentPatch.SetLabels(labels)
+
+	// Get the desired annotations and append them to the parent
+	desiredAnnotations, returnNamespaces, returnGKs := a.desiredParentAnnotations(mode == updateToSuperset)
+	annotations := a.parent.GetAnnotations()
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+	for k, v := range desiredAnnotations {
+		annotations[k] = v
+	}
+	parentPatch.SetAnnotations(annotations)
+
+	options := metav1.ApplyOptions{
+		FieldManager: a.fieldManager + "parent-labeller",
+		Force:        false,
+	}
+
+	// update parent in the cluster.
+	if !reflect.DeepEqual(original.GetLabels(), parentPatch.GetLabels()) ||
+		!reflect.DeepEqual(original.GetAnnotations(), parentPatch.GetAnnotations()) {
+		if _, err := a.parentClient.Apply(ctx, a.parent.GetName(), parentPatch, options); err != nil {
```

Comment:
this is an interesting one. our parent is our instance (https://github.com/kro-run/kro/pull/561/files#diff-b51fa143abc50a3ef21525703711d6ab24b80395539ca3a6a9a4d3da24050435R196), and we watch the instance in our controller as well AFAIU. Will this accidentally trigger a reconciliation if we don't filter for the annotations in our event handler? or potentially cause an early requeue?

Reply:
Naively i think it should trigger a watch change. How do you filter for annotations ? We dont usually know what field changed right ? 

Maybe it is ok. This happends during create and should "settle down" into a stable state with new reconciles being no-op (no change). 

Thoughts ?

Reply:
Actually I agree, the stable state and eventual consistency should win here. +1